### PR TITLE
Fix crash in some cases when count of labels more than `label.max_count`

### LINKED ISSUE #6597: Crash when count of labels more than label.max_count
**Describe the bug (REQUIRED)**
Game crashes when count of labels more than `label.max_count` 

**To Reproduce (REQUIRED)**

**Expected behavior (REQUIRED)**
the game shouldn't crash

**Defold version (REQUIRED):**
  - Version 1.2.192

**Platforms (REQUIRED):**
 - Platforms: Android

**Minimal repro case project (OPTIONAL):**
in progress.

**Logs (OPTIONAL):**
```
backtrace:
  #00  pc 000000000010972c  (dmGameObject::DoDeleteInstance(dmGameObject::Collection*, dmGameObject::Instance*)) (SourceCode: /home/runner/work/defold/defold/engine/gameobject/build/../src/gameobject/gameobject.cpp:1875)
  #00  pc 0000000000107a20  (dmGameObject::PostUpdate(dmGameObject::Collection*)) (SourceCode: /home/runner/work/defold/defold/engine/gameobject/build/../src/gameobject/gameobject.cpp:2621)
  #00  pc 0000000000107a20  (dmGameObject::PostUpdate(dmGameObject::CollectionHandle*)) (SourceCode: /home/runner/work/defold/defold/engine/gameobject/build/../src/gameobject/gameobject.cpp:2641)
  #00  pc 000000000012c17c  (dmGameSystem::CompCollectionProxyPostUpdate(dmGameObject::ComponentsPostUpdateParams const&)) (SourceCode: /home/runner/work/defold/defold/engine/gamesys/build/../src/gamesys/components/comp_collection_proxy.cpp:336)
  #00  pc 00000000001078a4  (dmGameObject::PostUpdate(dmGameObject::Collection*)) (SourceCode: /home/runner/work/defold/defold/engine/gameobject/build/../src/gameobject/gameobject.cpp:2576)
  #00  pc 00000000001078a4  (dmGameObject::PostUpdate(dmGameObject::CollectionHandle*)) (SourceCode: /home/runner/work/defold/defold/engine/gameobject/build/../src/gameobject/gameobject.cpp:2641)
  #00  pc 000000000012c17c  (dmGameSystem::CompCollectionProxyPostUpdate(dmGameObject::ComponentsPostUpdateParams const&)) (SourceCode: /home/runner/work/defold/defold/engine/gamesys/build/../src/gamesys/components/comp_collection_proxy.cpp:336)
  #00  pc 00000000001078a4  (dmGameObject::PostUpdate(dmGameObject::Collection*)) (SourceCode: /home/runner/work/defold/defold/engine/gameobject/build/../src/gameobject/gameobject.cpp:2576)
  #00  pc 00000000001078a4  (dmGameObject::PostUpdate(dmGameObject::CollectionHandle*)) (SourceCode: /home/runner/work/defold/defold/engine/gameobject/build/../src/gameobject/gameobject.cpp:2641)
  #00  pc 00000000000b4454  (dmEngine::Step(dmEngine::Engine*)) (SourceCode: /home/runner/work/defold/defold/engine/engine/build/../src/engine.cpp:1584)
  #00  pc 00000000000b517c  (dmEngineUpdate(dmEngine::Engine*)) (SourceCode: /home/runner/work/defold/defold/engine/engine/build/../src/engine.cpp:1998)
  #00  pc 00000000000b1370  (dmEngine::RunLoop(dmEngine::RunLoopParams const*)) (SourceCode: /home/runner/work/defold/defold/engine/engine/build/../src/engine_loop.cpp:69)
  #00  pc 00000000000b12f4  (EngineMain(int, char**)) (SourceCode: /home/runner/work/defold/defold/engine/engine/build/../src/engine_main.cpp:49)
  #00  pc 00000000000b12f4  (EngineMainThread(void*)) (SourceCode: /home/runner/work/defold/defold/engine/engine/build/../src/engine_main.cpp:68)
  #00  pc 0000000000219a8c  (dmThread::ThreadStartProxy(void*)) (SourceCode: /home/runner/work/defold/defold/engine/dlib/build/../src/dlib/thread.cpp:42)
  #00  pc 00000000000b6e44  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+264)
  #00  pc 0000000000053454  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+68)
```

**Workaround (OPTIONAL):**
Increase `label.max_count` in the `game.project`

### PR DESCRIPTION
**Pull request by AGulev**
Fixes crash that reproduces with the following conditions:
- objects spawned by a collection factory;
- objects has a parent-child hierarchy;
- the parent object has a component that can't be spawned because of max components limit.
 
Fix https://github.com/defold/defold/issues/6597

### CODE REVIEW

File: engine/gameobject/src/gameobject/gameobject.cpp [hunk 1]
```diff
@@ -1412,6 +1416,10 @@ namespace dmGameObject
                         ++component_instance_data_index;
                 }
             } else {
+                ReparentChildNodes(collection, instance);
+                Unlink(collection, instance);
+                MoveAllUp(collection, instance);
+
```
Comment (AGulev) [inline engine/gameobject/src/gameobject/gameobject.cpp:1422]: Proper reparenting of the nodes before removing

File: engine/gameobject/src/gameobject/gameobject.cpp [hunk 2]
```diff
@@ -1831,34 +1839,8 @@ namespace dmGameObject
         instance->m_ToBeAdded = 0;
     }
 
-    static void DoDeleteInstance(Collection* collection, HInstance instance)
```
Comment (AGulev) [inline engine/gameobject/src/gameobject/gameobject.cpp:1834]: I just moved a part of DoDeleteInstance related to re-parenting into the other method to be able to call it from two different places

File: engine/gameobject/src/gameobject/gameobject.cpp [hunk 3]
```diff
@@ -1141,6 +1141,10 @@ namespace dmGameObject
         return instance;
     }
 
+    static void Unlink(Collection* collection, Instance* instance);
+    static void ReparentChildNodes(Collection* collection, Instance* instance);
+    static void MoveAllUp(Collection* collection, Instance* instance);
```
Comment (AGulev) [inline engine/gameobject/src/gameobject/gameobject.cpp:1146]: not sure if I should move a lot of code from one place to another. This looks a bit more clear from diff point of view, but not sure it it's fine.

Comment (britzl) [reply engine/gameobject/src/gameobject/gameobject.cpp:1146]: Yeah, not sure either. Forward declarations like this are not very common in our codebase.

**Review by britzl — DISMISSED** [review]
Looks good. Minor question about the forward declarations. Perhaps as @Jhonnyg opinion?

**Comment by AGulev** [comment]
@britzl one more thing : I decided to add a test for that case, it's not so easy but it's pretty important, I think

**Comment by britzl** [comment]
> I decided to add a test for that case

Very wise!

**Pull request merged**